File: kombi-lib/src/main/java/com/sgnatiuk/cartesian/CartesianProductMap.java
package com.sgnatiuk.cartesian;

import java.io.Serializable;
import java.util.*;
import java.util.stream.Collectors;

class CartesianProductMap<K, V> extends EncodableCartesianProduct<Map<K, V>> implements Serializable {

    private final Map<K, Object[]> values;
    private final Object[] dataKeys;
    private final boolean keepOrder;

    CartesianProductMap(Map<K, ? extends Collection<V>> values, boolean keepOrder) {
        this.values = copyWithOrder(
                values,
                keepOrder ? (o1, o2) -> 0 : new ValuesCountDesc<>()
        );
        this.dataKeys = this.values.keySet().toArray();
        this.keepOrder = keepOrder;
    }

    private CartesianProductMap(boolean keepOrder, Map<K, Object[]> values) {
        this.values = values;
        this.dataKeys = this.values.keySet().toArray();
        this.keepOrder = keepOrder;
    }

    private Map<K, Object[]> copyWithOrder(
            Map<K, ? extends Collection<V>> data,
            Comparator<Map.Entry<K, ? extends Collection<V>>> comparator
    ) {
        return data.entrySet()
                .stream()
                .sorted(comparator)
                .map(entry -> new AbstractMap.SimpleEntry<>(
                        entry.getKey(),
                        new ArrayList<>(entry.getValue())
                ))
                .collect(Collectors.toMap(
                        AbstractMap.SimpleEntry::getKey,
                        pair -> pair.getValue().toArray(),
                        (v1, v2) -> {
                            throw new IllegalArgumentException("Unexpected key duplication in data:" + data);
                        },
                        LinkedHashMap::new
                ));
    }

    @SuppressWarnings("unchecked")
    @Override
    protected MaskDecoder<Map<K, V>> maskDecoder() {
        return encoded -> {
            HashMap<K, V> decoded = new HashMap<>(encoded.length);
            for (int i = 0; i < encoded.length; i++) {
                K fieldKey = (K) dataKeys[i];
                V value = (V) values.get(fieldKey)[encoded[i]];
                decoded.put(fieldKey, value);
            }
            return decoded;
        };
    }

    @Override
    protected Object[][] values() {
        Object[][] data = new Object[values.size()][];
        int index = 0;
        for (Object[] value : values.values()) {
            data[index++] = value;
        }
        return data;
    }

    @SuppressWarnings("unchecked")
    @Override
    public List<CartesianProduct<Map<K, V>>> split(int n) {
        if (n < 2) {
            return Collections.singletonList(this);
        }
        List<CartesianProduct<Map<K, V>>> splitList = new ArrayList<>(n);

        K keyOfMaxLengthArr = (K) dataKeys[0];
        Object[] arrWithMaxLength = values.get(keyOfMaxLengthArr);
        for (int i = 1; i < dataKeys.length; i++) {
            Object[] nextArr = values.get(dataKeys[i]);
            if (nextArr.length > arrWithMaxLength.length) {
                keyOfMaxLengthArr = (K) dataKeys[i];
                arrWithMaxLength = nextArr;
            }
        }
        int parts = Math.min(n, arrWithMaxLength.length);

        int from = 0;
        for (int i = 0; i < parts; i++) {
            int valuesPerChunk = (arrWithMaxLength.length - from) / (parts - i);
            int to = from + valuesPerChunk;

            Map<K, Object[]> nd = new HashMap<>(values);
            nd.put(
                    keyOfMaxLengthArr,
                    Arrays.copyOfRange(arrWithMaxLength, from, to)
            );
            splitList.add(new CartesianProductMap<>(keepOrder, nd));
            from = to;
        }
        return splitList;
    }


    private static class ValuesCountDesc<K, V> implements Comparator<Map.Entry<K, ? extends Collection<V>>> {

        @Override
        public int compare(
                Map.Entry<K, ? extends Collection<V>> o1,
                Map.Entry<K, ? extends Collection<V>> o2
        ) {
            return o2.getValue().size() - o1.getValue().size();
        }
    }
}

File: kombi-lib/src/main/kotlin/com/sgnatiuk/cartesian/CartesianProductMap.kt
package com.sgnatiuk.cartesian

import com.sgnatiuk.Splittable
import com.sgnatiuk.cartesian.encodable.EncodableCartesianProduct
import com.sgnatiuk.cartesian.encodable.decoders.MaskDecoderMap
import kotlin.collections.Map.Entry

internal class CartesianProductMap<K, V> (
        data: Map<K, Collection<V>>,
        keepOrder: Boolean = false
) : EncodableCartesianProduct<Map<K, V>>(), Splittable<CartesianProductMap<K, V>> {


    private val internalData: Map<K, List<V>> = convertToFixedOrderMap(data, keepOrder)
    private val dataKeys = ArrayList<K>(internalData.keys)

    override val decoder = MaskDecoderMap(internalData, dataKeys)
    override val values
        get() = internalData.values


    private fun convertToFixedOrderMap(
            data: Map<K, Collection<V>>,
            keepOrder: Boolean
    ): Map<K, List<V>>  = if (keepOrder) {
        data.mapValues {
            ArrayList(it.value)
        }
    } else {
        sortByValuesCount(data)
    }

    private fun sortByValuesCount(
            data: Map<K, Collection<V>>,
            comparator: Comparator<Map.Entry<K, Collection<V>>> = ValuesCountAsc()
    ) : Map<K, List<V>> {
        val keepOrderMap: MutableMap<K, List<V>> = LinkedHashMap()
        data.entries.sortedWith(comparator).forEach {
            keepOrderMap[it.key] = ArrayList(it.value)
        }
        return keepOrderMap
    }

    //implement more fair split
    override fun split(n: Int): List<CartesianProductMap<K, V>> {
        val coders = ArrayList<CartesianProductMap<K, V>>()
        val descSortedData = sortByValuesCount(
                internalData,
                ValuesCountAsc<K, V>().reversed()
        )
        val firstFieldEntry = descSortedData.entries.first()
        val firstFieldValues = firstFieldEntry.value
        val parts = Math.min(n, firstFieldValues.size)

        var from = 0
        repeat(parts){ i ->
            val fieldValuesPerChunk = (firstFieldValues.size - from) / (parts - i)
            val to = from + fieldValuesPerChunk
            val newData = LinkedHashMap(descSortedData)
            newData[firstFieldEntry.key] = firstFieldValues.subList(from, to)
            coders += CartesianProductMap(newData)
            from = to
        }

        return coders
    }

    class ValuesCountAsc<K, T> : Comparator<Map.Entry<K, Collection<T>>>{

        override fun compare(
                o1: Entry<K, Collection<T>>,
                o2: Entry<K, Collection<T>>
        ) : Int = o1.value.size - o2.value.size
    }
}